test/torchaudio_unittest/functional/functional_impl.py

```diff
@@ -106,6 +106,84 @@ def test_lfilter_9th_order_filter_stability(self):
+        corresponding to a pure delay filter imparts no time delay.
+        """
+
+        torch.random.manual_seed(42)
+        waveform = torch.rand(2, 44100 * 1, dtype=self.dtype, device=self.device)
+        b_coeffs = torch.tensor([0, 0, 0, 1], dtype=self.dtype, device=self.device)
+        a_coeffs = torch.tensor([1, 0, 0, 0], dtype=self.dtype, device=self.device)
+        padded_waveform = torch.cat((waveform, torch.zeros(2, 3, dtype=self.dtype, device=self.device)), axis=1)
+        output_waveform = F.filtfilt(padded_waveform, a_coeffs, b_coeffs)
+
+        self.assertEqual(output_waveform, padded_waveform, atol=1e-5, rtol=1e-5)
+
+    def test_filtfilt_filter_sinusoid(self):
+        """
+        Check that, for a signal comprising two sinusoids, applying filtfilt
+        with appropriate filter coefficients correctly removes the higher-frequency
+        sinusoid while imparting no time delay.
+        """
+
+        T = 1.0
+        samples = 1000
+
+        torch.random.manual_seed(42)
+        timesteps = torch.linspace(0, T, samples, dtype=self.dtype, device=self.device)
+        k0 = 10
+        k1 = 200
+        waveform_k0 = torch.cos(2 * torch.pi * k0 * timesteps).to(self.dtype).to(self.device)
+        waveform_k1 = torch.cos(2 * torch.pi * k1 * timesteps).to(self.dtype).to(self.device)
+        waveform = waveform_k0 + waveform_k1
+
+        # Transfer function numerator and denominator polynomial coefficients
+        # corresponding to 8th-order Butterworth filter with 100-cycle/T cutoff.
+        # Generated with
+        # >>> from scipy import signal
+        # >>> b_coeffs, a_coeffs = signal.butter(8, 0.2)
+        b_coeffs = torch.tensor(
+            [
+                2.39596441e-05,
+                1.91677153e-04,
+                6.70870035e-04,
+                1.34174007e-03,
+                1.67717509e-03,
+                1.34174007e-03,
+                6.70870035e-04,
+                1.91677153e-04,
+                2.39596441e-05,
+            ],
+            dtype=self.dtype,
+            device=self.device,
+        )
+        a_coeffs = torch.tensor(
+            [
+                1.,
+                -4.78451489,
+                10.44504107,
+                -13.45771989,
+                11.12933104,
+                -6.0252604,
+                2.0792738,
+                -0.41721716,
+                0.0372001,
+            ],
+            dtype=self.dtype,
+            device=self.device,
+        )
+
+        # Extend waveform in each direction, preserving periodicity.
+        padded_waveform = torch.cat((waveform[:-1], waveform, waveform[1:]))
+
+        output_waveform = F.filtfilt(padded_waveform, a_coeffs, b_coeffs)
+
+        # Remove padding from output waveform; confirm that result
+        # closely matches waveform_k0.
+        self.assertEqual(output_waveform[samples - 1: 2 * samples - 1], waveform_k0, atol=1e-3, rtol=1e-3)
```

Comment:
nit: ``output_waveform[samples - 1 : 2 * samples - 1]``

Could ``atol`` and ``rtol`` be smaller than 1e-3, or it will throw an error if we decrease it?

Reply:
the linter likes it without the space

the maximum error is around 1e-4

Reply:
let's try it with 1e-4 to see if it passes the test. As ``lfilter`` has a lower threshold (1e-5) for testing.

Reply:
it fails with 1e-4, unfortunately